### ISSUE TITLE
JRE 9 - optifine causes failure to boot

### ISSUE THREAD

**Issue by sylcai** [sp614x/optifine#192]
```
Exception in thread "main" java.lang.ClassCastException: jdk.internal.loader.ClassLoaders$AppClassLoader (in module: java.base) cannot be cast to java.net.URLClassLoader (in module: java.base)
    at net.minecraft.launchwrapper.Launch.<init>(Launch.java:33)
    at net.minecraft.launchwrapper.Launch.main(Launch.java:27)

```

Optifine has issues running on JRE 9 - I know its in development, but it's still annoying, as JRE 9 has some optimizations that increase my framerate substantially with minecraft.

**Comment by ajthemacboy** [comment]
Confirmed. I too would really like to use 9 for the increased performance combined with OptiFine.

**Comment by SetantaLP** [comment]
As far as i know, this is a problem inside Mojang's launchwrapper that Optifine uses. Not inside Optifine itself. So fixing the error in Optifine would only work, if the launchwrapper gets an update.

**Comment by GreyAsteroid** [comment]
Since this issue was filed, JRE 9 has become feature complete and is now less than 6 months from release. While vanilla minecraft runs perfectly fine with JRE 9, Optifine on the other hand crashes almost instantly.

**Comment by SetantaLP** [comment]
@GreyAsteroid It's not optifine itself that crashs. It's [mojang's LegacyLauncher](https://github.com/Mojang/LegacyLauncher) that crashs, because of a change to the class loader in java 9. And it seems that this issue might not be fixed (see Mojang/LegacyLauncher#11), because java 9 modularization is a very tricky thing.
So there are two ways to solve this issue here:
1. If the optifine installer detects a java9 environment, it downloads a modified version of the legacy launcher (or creates a copy of the launchwrapper.jar and applies a patch), that fixes this problem as suggested in the pull request mentioned above.
2. Optifine stops using the launch wrapper and creates a modified jar while installing.

**Comment by Brottweiler** [comment]
JRE 9 is released now, and this is an issue.

**Comment by SetantaLP** [comment]
@Brottweiler right, but that's not a bug in optifine (see my posts above), so optifine is not responsible for fixing that. And besides that, the bug only affects users who manually change the jre settings in the launcher to java 9. All other players are not affected, either because they use the mojang-jre (which will stay on java 8, as long as the launchwrapper is incompatible with java 9), or because they use java 8 or below.

**Comment by sp614x** [comment]
This is a problem in the Launchwrapper library that the launcher uses to inject mods in the classpath.
Not an OptiFine problem.

**Comment by Orakul-tomataus** [comment]
Does not launch with Java 11 (

**Comment by SetantaLP** [comment]
@Orakul-tomataus Yea, but that's most likely the same bug in the launch wrapper library mentioned above and not a bug inside optifine itself.

**Comment by Orakul-tomataus** [comment]
@SetantaLP Thanks, I already put the bones (java 8). )))

**Comment by sp614x** [comment]
Updated launchwrapper by Runemoro, should be compatible with Java 8, 9, 10:
https://github.com/sp614x/optifine/issues/1739#issuecomment-417884419

**Comment by sp614x** [comment]
Fixed installer to install launchwrapper-2.0.
All new versions would use the new installer.
